File: src/0902/S0902_v1.java
class Solution {
    public int atMostNGivenDigitSet(String[] digits, int n) {
        String strn = String.valueOf(n);
        int nlen = strn.length();
        int dlen = digits.length;
        int res = 0;
        int[] plist = new int[nlen];
        plist[nlen-1] = 1;
        if (nlen > 1) {
            plist[nlen-2] = dlen;
            res += dlen;
            if (nlen > 2) {
                for (int i=nlen-3;i>=0;i--) {
                    plist[i] = plist[i+1] * dlen;
                    res += plist[i];
                }
            }
        }
        int i = 0, check = 0, opt= 0;
        for (i=0;i<nlen;i++) {
            check = 0; opt = 0;
            for (String s : digits) {
                if (Integer.valueOf(s) <= Integer.valueOf(strn.charAt(i))-48) check++;
                if (Integer.valueOf(s) < Integer.valueOf(strn.charAt(i))-48) opt++;
            }
            if (check == 0) break;
            res += plist[i] * opt;
            if (check == opt) break;
        }
        return (i == nlen && check != 0) ? res + check - opt : res;
    }
}

File: src/0902/S0902_v1.py
class Solution:
    def atMostNGivenDigitSet(self, digits: List[str], n: int) -> int:
        strn = str(n)
        nlen = len(strn)
        dlen = len(digits)
        plist = []
        for i in range(nlen-1, 0, -1):
            plist.append(dlen**i)
        res = sum(plist)
        plist.append(1)
        i, check, opt = 0, 0, 0
        for i, v in enumerate(strn):
            check = sum(int(j)*0+1 for j in digits if int(j) <= int(v))
            if check == 0:
                break
            opt = sum(int(j)*0+1 for j in digits if int(j) < int(v))
            res += plist[i] * opt
            if check == opt:
                break
        print(plist)
        return res + check - opt if i == nlen - 1 and check != 0 else res
